version: "3.7"

x-logging: &default-logging
  driver: local
  options:
    max-size: "10m"
    max-file: "3"

x-healthcheck: &default-health-check
  interval: 30s
  timeout: 10s
  retries: 5

services:
  app:
    build:
      context: ./
      target: runner
    ports:
      - "3000:3000"
    volumes:
      - ./src:/app/src:rw,cached
      - ./public:/app/public:rw,cached
      - ./package-lock.json:/app/package-lock.json:rw,cached
      - ./package.json:/app/package.json:rw,cached
      - ./tsconfig.json:/app/tsconfig.json:rw,cached
      - npm_cache:/root/.npm:rw,cached
      - node_modules:/app/node_modules

  caddy:
    build:
      context: ./docker/caddy
      target: local
    depends_on:
      - app
    environment:
      DOMAIN: ${DOMAIN}
    restart: unless-stopped
    volumes:
      - ./:/app:ro
      - ./docker/caddy/Caddyfile.prod:/etc/caddy/Caddyfile:ro
      - caddy_data:/data
      - caddy_config:/config
    ports:
      - "80:80"
      - "443:443"
    healthcheck: *default-health-check
    logging: *default-logging

volumes:
  caddy_data:
  caddy_config:
  npm_cache:
  node_modules:
